### PR TITLE
Always check Java pointers for null before deref

### PR DESCRIPTION
**Pull request by smowton**
This also improves goto-check's resolution for finding symbol modes by finding the function in the symbol table and parameterising from above.

I wrote this because the test generator was producing strange results due to missing null checks. Before the test generator became a separate module it twiddled the command-line options to add `pointer-check`; however I think doing it this way is better, as it is never semantically sane to analyse Java-but-null-pointers-are-wild.

### CODE REVIEW

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -1877,7 +1873,11 @@ void goto_check(
 
   Forall_goto_functions(it, goto_functions)
   {
-    goto_check.goto_check(it->second);
+    const symbolt *sym;
+    irep_idt mode;
+    if(!ns.lookup(it->first, sym))
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Shouldn't it be an invariant that the symbol exists? I think it should be safe to use
```
irep_idt mode=ns.lookup(it->first).mode;
```
?

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -993,9 +992,10 @@ void goto_checkt::pointer_validity_check(
   const dereference_exprt &expr,
   const guardt &guard,
   const exprt &access_lb,
-  const exprt &access_ub)
+  const exprt &access_ub,
+  const irep_idt &mode)
 {
-  if(!enable_pointer_check)
+  if(mode!=ID_java && !enable_pointer_check)
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: So we would always be doing this for `mode==ID_java`, whether it's enabled or not!?

Comment (smowton) [inline src/analyses/goto_check.cpp]: Yes. Check-deref'd-pointer-not-null is mandatory for Java.

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (smowton) [inline src/analyses/goto_check.cpp]: Unsure, hence being careful.

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: But we should then be generating an exception, not an assertion, shouldn't we?

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: I'd go all-in and wait for bug reports (or failing regression tests).

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (smowton) [inline src/analyses/goto_check.cpp]: Ideally yes. However this matches current Java front-end behaviour for out of bounds array access and similar. I think we should support this simple model and then perhaps in the future add support for a "full fidelity" model where we support catching NPE, AOOBE and so on.

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (smowton) [inline src/analyses/goto_check.cpp]: Ok will change this shortly

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Ok; I'd thus suggestion that the help output for --pointer-check be amended to say "C/C++ only".

Comment (smowton) [inline src/analyses/goto_check.cpp]: Done

**Pull request merged**